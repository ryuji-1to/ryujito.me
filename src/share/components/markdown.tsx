type Props = {
  children: NonNullable<
    React.ComponentProps<"div">["dangerouslySetInnerHTML"]
  >["__html"];
};

function replaceSmall(htmlText: string) {
  const cardStartTag = "<p>--small:start--</p>";
  const cardEndTag = "<p>--small:end--</p>";
  const cardStartDiv = `<div style='font-size:12px !important;'>
    `;
  const cardEndDiv = "</div>";
  if (!htmlText.includes(cardStartTag) || !htmlText.includes(cardEndTag)) {
    return htmlText;
  }
  return htmlText
    .replaceAll(cardStartTag, cardStartDiv)
    .replaceAll(cardEndTag, cardEndDiv);
}

function replaceCard(htmlText: string) {
  const cardStartTag = "<p>--card:start--</p>";
  const cardEndTag = "<p>--card:end--</p>";
  const cardStartDiv =
    "<div class='bg-gray-100 dark:bg-gray-900 border border-gray-300 dark:border-gray-700 py-2 px-5 my-2 rounded-lg'>";
  const cardEndDiv = "</div>";
  if (!htmlText.includes(cardStartTag) || !htmlText.includes(cardEndTag)) {
    return htmlText;
  }
  return htmlText
    .replaceAll(cardStartTag, cardStartDiv)
    .replaceAll(cardEndTag, cardEndDiv);
}

function replaceMark(htmlText: string) {
  // const mark = "<p>--mark:print--</p>";
  // if (!htmlText.includes(mark)) {
  //   return htmlText;
  // }
  // const printStyle = `<div class='print-only' style='font-size: 10px; text-align: right;'>â€» This resume is generated from https://ryujito.me/resume.</div>`;
  // return htmlText.replaceAll(mark, printStyle);
  return htmlText;
}

function replaceBrowser(htmlText: string) {
  const browserStart = "<p>--browser:start--</p>";
  const browserEnd = "<p>--browser:end--</p>";
  const start = `
  <div class="bg-gray-50 dark:bg-gray-800 rounded-sm drop-shadow-lg min-h-20">
     <div class="h-6 bg-gray-200 dark:bg-gray-700 py-1 px-2 rounded-t">
        <div class="flex items-center h-full gap-1.5">
            <div class="rounded-full bg-red-500 h-2 w-2"></div>
            <div class="rounded-full bg-amber-500 h-2 w-2"></div>
            <div class="rounded-full bg-green-500 h-2 w-2"></div>
        </div>
      </div>
      <div class="px-4 py-1">
  `;
  const end = "</div></div>";
  return htmlText.replaceAll(browserStart, start).replaceAll(browserEnd, end);
}

function replaceHtml(
  htmlText: string,
  methods: ((htmlText: string) => string)[],
) {
  return methods.reduce(
    (modifiedText, transform) => transform(modifiedText),
    htmlText,
  );
}

export function Markdown(props: Props) {
  const value = props.children.toString();
  const html = replaceHtml(value, [
    replaceCard,
    replaceSmall,
    replaceMark,
    replaceBrowser,
  ]);
  return (
    <div
      className="prose prose-md dark:prose-invert prose-h1:text-balance"
      // biome-ignore lint/security/noDangerouslySetInnerHtml: <explanation>
      dangerouslySetInnerHTML={{ __html: html }}
    />
  );
}
