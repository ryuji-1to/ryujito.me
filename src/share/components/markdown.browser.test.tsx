import { render } from "@testing-library/react";
import { Markdown } from "./markdown";

function wrapMarkdownRoot(htmlText: string) {
  return `
    <div class="prose prose-md dark:prose-invert prose-h1:text-balance">
    ${htmlText}
    </div>
    `;
}

function trimSpace(v: string) {
  return v.replace(/\s+/g, "");
}

function formatExpected(expected: string) {
  return trimSpace(wrapMarkdownRoot(expected));
}

describe("share/components/markdown", () => {
  test("snapshot", () => {
    const { getByText } = render(<Markdown>hoge</Markdown>);
    const target = getByText("hoge");
    expect(target).toMatchSnapshot();
  });
  test("replaceSmall", () => {
    const actual = `
    <p>--small:start--</p>
    <div>test</div>
    <p>--small:end--</p>
    `;
    const expected = `
    <div class="text-xs">
    <div>test</div>
    </div>
    `;

    const { container } = render(<Markdown>{actual}</Markdown>);
    expect(trimSpace(container.innerHTML.toString())).toMatch(
      formatExpected(expected),
    );
  });
  test("replaceCard", () => {
    const actual = `
    <p>--card:start--</p>
    <div>test</div>
    <p>--card:end--</p>
    `;
    const expected = `
    <div class="bg-gray-2 dark:bg-dark-gray-2 border border-gray-5 dark:border-dark-gray-5 py-8 px-24 my-16 rounded-8">
    <div>test</div>
    </div>
    `;
    const { container } = render(<Markdown>{actual}</Markdown>);
    expect(trimSpace(container.innerHTML.toString())).toMatch(
      formatExpected(expected),
    );
  });
  test("replaceBrowser", () => {
    const actual = `
    <p>--browser:start--</p>
    <div>test</div>
    <p>--browser:end--</p>
    `;
    const expected = `
    <div class="bg-gray-1 dark:bg-dark-gray-1 border border-gray-4 dark:border-dark-gray-4 rounded-4 drop-shadow-lg min-h-64">
      <div class="h-24 bg-gray-4 dark:bg-dark-gray-4 py-4 px-8 rounded-t-4">
        <div class="flex items-center h-full gap-4">
            <div class="rounded-full bg-red dark:bg-dark-red h-8 w-8"></div>
            <div class="rounded-full bg-orange dark:bg-dark-orange h-8 w-8"></div>
            <div class="rounded-full bg-green dark:bg-dark-green h-8 w-8"></div>
        </div>
      </div>
      <div class="px-8 py-4">
        <div>test</div>
      </div>
    </div>
    `;
    const { container } = render(<Markdown>{actual}</Markdown>);
    expect(trimSpace(container.innerHTML.toString())).toMatch(
      formatExpected(expected),
    );
  });
});
