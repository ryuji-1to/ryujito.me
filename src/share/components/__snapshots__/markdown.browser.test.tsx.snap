// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`share/components/markdown > snapshot 1`] = `
<div
  class="prose prose-md dark:prose-invert prose-h1:text-balance"
>
  hoge
</div>
`;
